sdks/java/core/src/main/java/org/apache/beam/sdk/io/TFRecordIO.java

```diff
@@ -0,0 +1,878 @@
+    @VisibleForTesting
+    TFRecordSink(ValueProvider<String> baseOutputFilename,
+                 String extension,
+                 String fileNameTemplate,
+                 TFRecordIO.CompressionType compressionType) {
+      super(baseOutputFilename, extension, fileNameTemplate,
+          writableByteChannelFactory(compressionType));
+    }
+
+    @Override
+    public FileBasedWriteOperation<byte[]> createWriteOperation(PipelineOptions options) {
+      return new TFRecordWriteOperation(this);
+    }
+
+    private static WritableByteChannelFactory writableByteChannelFactory(
+        TFRecordIO.CompressionType compressionType) {
+      switch (compressionType) {
+        case AUTO:
+          throw new IllegalArgumentException("Unsupported compression type AUTO");
+        case NONE:
+          return CompressionType.UNCOMPRESSED;
+        case GZIP:
+          return CompressionType.GZIP;
+        case DEFLATE:
+          return CompressionType.DEFLATE;
+      }
+      return CompressionType.UNCOMPRESSED;
+    }
+
+    /**
+     * A {@link org.apache.beam.sdk.io.FileBasedSink.FileBasedWriteOperation
+     * FileBasedWriteOperation} for TFRecord files.
+     */
+    private static class TFRecordWriteOperation extends FileBasedWriteOperation<byte[]> {
+      private TFRecordWriteOperation(TFRecordSink sink) {
+        super(sink);
+      }
+
+      @Override
+      public FileBasedWriter<byte[]> createWriter(PipelineOptions options) throws Exception {
+        return new TFRecordWriter(this);
+      }
+    }
+
+    /**
+     * A {@link org.apache.beam.sdk.io.FileBasedSink.FileBasedWriter FileBasedWriter}
+     * for TFRecord files.
+     */
+    private static class TFRecordWriter extends FileBasedWriter<byte[]> {
+      private WritableByteChannel outChannel;
+      private TFRecordCodec codec;
+
+      private TFRecordWriter(FileBasedWriteOperation<byte[]> writeOperation) {
+        super(writeOperation);
+        this.mimeType = MimeTypes.BINARY;
+      }
+
+      @Override
+      protected void prepareWrite(WritableByteChannel channel) throws Exception {
+        this.outChannel = channel;
+        this.codec = new TFRecordCodec();
+      }
+
+      @Override
+      public void write(byte[] value) throws Exception {
+        codec.write(outChannel, value);
+      }
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+
+  private static class TFRecordCodec {
+    private static final int HEADER_LEN = (Long.SIZE + Integer.SIZE) / Byte.SIZE;
+    private static final int FOOTER_LEN = Integer.SIZE / Byte.SIZE;
+    private static HashFunction crc32c = Hashing.crc32c();
+
+    private ByteBuffer header = ByteBuffer.allocate(HEADER_LEN).order(ByteOrder.LITTLE_ENDIAN);
+    private ByteBuffer footer = ByteBuffer.allocate(FOOTER_LEN).order(ByteOrder.LITTLE_ENDIAN);
+
+    private int mask(int crc) {
+      return ((crc >>> 15) | (crc << 17)) + 0xa282ead8;
```

Comment:
The number 0xffffffff is not a good example, since the result is the same for signed/unsigned shift.

What is the result of 0x80000000 << 1?
Unsigned shift will make it 0, while signed shift will make it 0x80000000.
0xa0000000 << 1 would be 0x40000000 with unsigned shift, but 0xc0000000 with signed shift.
It's a subtle difference and you'll end up the situation that lots of records saved in Java version would be treated as corrupted in C++/python readers.